052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,156 @@
+Today ZooKeeper also uses jmx_exporter to expose its metrics. ZooKeeper is going away and will be replaced by KRaft controllers which also support Kafka metrics reporters.
+
+## Affected/not affected projects
+
+This affects `strimzi-kafka-operator `. Since ZooKeeper does not support metrics reporters, we should keep the current mechanism using jmx_exporter while we support ZooKeeper deployments.
+
+The reporter will be a new project and should be usable without Strimzi for both brokers and client side applications. As it's new code, it should be first release as a preview feature before becoming the preferred mechanism once KRaft is the default.
+
+In `strimzi-kafka-operator`, we will need new examples demonstrating how to use the reporter, new grafana dashboards with metric names from the reporter.
+
+## Compatibility
+
+Differences with jmx_exporter metrics:
+
+- The reporter does not compute 1/5/15 minute rate, mean, max, min, stddev metrics. It's preferable to compute them in Prometheus instead of on the client side.
+- The reporter is missing the `kafka_server_app_info_starttimems` metric with the client/broker id label. (Due to [KAFKA-15186](https://issues.apache.org/jira/browse/KAFKA-15186))
+
+Assuming jmx_exporter does not have any rules, this is the other main metric change:
+
+- With the reporter, the `name` field is put directly in the metric name. For example this MBean, `kafka.server:type=ZooKeeperClientMetrics,name=ZooKeeperRequestLatencyMs` with the `Count` attribute is converted to `kafka_server_zookeeperclientmetrics_zookeeperrequestlatencyms_count`. By default jmx_exporter keeps the name as a label, `kafka_server_zookeeperclientmetrics_count{name="ZooKeeperRequestLatencyMs",}`.
+
+### Comparison
+
+Actually with the example rules from [kafka-metrics.yaml](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/kafka-metrics.yaml), jmx_exporter generates metric names very similar to the reporter.
+
+See https://gist.github.com/mimaison/335bc52bd5fb39097b7e6865c8cd1bea that contains the output from jmx_exporter with the default [metrics.yaml](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/kafka-metrics.yaml) rules, and the proposed metrics reporter with the following configuration.
+
+```
+prometheus.metrics.reporter.allowlist=kafka_cluster.*,kafka_controller.*,kafka_log.*,kafka_network.*,kafka_server_(brokertopicmetrics|delayedoperationpurgatory|fetchsessioncache|kafkarequesthandlerpool|kafkaserver|replicaalterlogdirsmanager|replicafetchermanager|replicamanager|sessionexpirelistener|socket_server_metrics|zookeeperclientmetrics).*,kafka_utils.*
+```
+
+Both files have been sorted and the comments have been removed so they are easy to compare.
+
+If we also remove the values, doing a diff highlights the following differences:
+
+- The JMX metrics only exist with jmx_exporter.
```

Comment:
I don't think we care about JMX metrics if we don't use JMX. But what about JVM metrics? Similarly, it would be good to cover how additional metrics can be added to this reporter. E.g.:
* Connector metrics in Connect
* OAuth metrics
* etc.

Reply:
Yes with reporters, JMX metrics are irrelevant, I just noted it to mention all the differences between the 2 outputs.

Regarding the client-side components it's tricky. Reporters are per Kafka client so in applications combining several clients (Connect, Streams, Bridge) it's not easy to setup, to the point where I would not recommend doing it in most cases. I'm trying to work upstream to see what could be done to address this.

For OAuth metrics, since it's a plugin, it's a completely different issue which I'm trying to tackle it in [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics).

Reply:
So, can you incorporate this into the proposals? Including the details about how to get JVM metrics, what does this mean for plugin metrics, what does this mean for Connect and other operands etc.?

Reply:
I pushed an updates which should clarify these points.